<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [93]:
import pandas as pd
import psycopg2

In [94]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [95]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [96]:
# текст запроса
query_3_1 = f'''select count(*)
                from vacancies 
                limit 10
'''

In [97]:
vacancies_count=pd.read_sql_query(query_3_1,connection)
vacancies_count

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\3747696216.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count=pd.read_sql_query(query_3_1,connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [98]:
# текст запроса
query_3_2 = f'''select count(distinct id)
                from employers
'''


In [99]:
# результат запроса
employers_count=pd.read_sql_query(query_3_2,connection)
employers_count

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\1747186731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_count=pd.read_sql_query(query_3_2,connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [100]:
# текст запроса
query_3_3 = f'''select count(distinct id)
                from areas
'''

In [101]:
# результат запроса
areas_count=pd.read_sql_query(query_3_3,connection)
areas_count

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\2574634389.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_count=pd.read_sql_query(query_3_3,connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [102]:
# текст запроса
query_3_4 = f'''select count(distinct id)
                from industries
'''

In [103]:
# результат запроса
industries_count=pd.read_sql_query(query_3_4,connection)
industries_count

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\2340521402.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_count=pd.read_sql_query(query_3_4,connection)


count
0    294

***

# выводы по предварительному анализу данных
Итого у нас получается окло 50тыс ваканский и 23,5 тыс работодателей в 1362 регионах.
Учитывая что не все работодатели ищут сотрудников и не во всех регионах можно предположить что очень грубо в среднем
каждому работодателю требуется от 2х до 3х сотрудников, более точные распределния посмотрим далее в графиках.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [104]:
# текст запроса
query_4_1 = f'''select a.name, count(v.name)
                from areas as a join vacancies as v on a.id=v.area_id
                group by 1
                order by 2 desc
'''

In [105]:
# результат запроса
cnt=pd.read_sql_query(query_4_1,connection)
cnt

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\1230476275.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt=pd.read_sql_query(query_4_1,connection)


name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [106]:
# текст запроса
query_4_2 = f'''select count(*)
                from vacancies
                where salary_from is not null or salary_to is not null
'''

In [107]:
# результат запроса
salary_is=pd.read_sql_query(query_4_2,connection)
salary_is

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\4020637221.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_is=pd.read_sql_query(query_4_2,connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [108]:
# текст запроса
query_4_3 = f'''select avg(salary_from) as low_salary ,avg(salary_to) high_salary
                from vacancies
'''

In [109]:
# результат запроса
salary_range=pd.read_sql_query(query_4_3,connection)
salary_range

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\421662210.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_range=pd.read_sql_query(query_4_3,connection)


low_salary    high_salary
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [110]:
# текст запроса
query_4_4 = f'''select schedule, employment, count(name) as vacancy_cnt
                from vacancies
                group by 1,2
                order by 3 desc
'''

In [111]:
# результат запроса
vacancy_cnt=pd.read_sql_query(query_4_4,connection)
vacancy_cnt

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\2476083013.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancy_cnt=pd.read_sql_query(query_4_4,connection)


schedule           employment  vacancy_cnt
0        Полный день     Полная занятость        35367
1   Удаленная работа     Полная занятость         7802
2      Гибкий график     Полная занятость         1593
3   Удаленная работа  Частичная занятость         1312
4     Сменный график     Полная занятость          940
5        Полный день           Стажировка          569
6     Вахтовый метод     Полная занятость          367
7        Полный день  Частичная занятость          347
8      Гибкий график  Частичная занятость          312
9        Полный день     Проектная работа          141
10  Удаленная работа     Проектная работа          133
11     Гибкий график           Стажировка          116
12    Сменный график  Частичная занятость          101
13  Удаленная работа           Стажировка           64
14     Гибкий график     Проектная работа           18
15    Сменный график           Стажировка           12
16    Вахтовый метод     Проектная работа            2
17    Сменный график     Проектная работа            1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [112]:
# текст запроса
query_4_5 = f'''select experience, count(name)
                from vacancies
                group by 1
                order by 2
'''

In [113]:
# результат запроса
experience=pd.read_sql_query(query_4_5,connection)
experience

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\1148048373.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience=pd.read_sql_query(query_4_5,connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
Как видим больше всего требуются сотрудники на полный рабочий день, с зарапотной платой от 70тыс руб до 110тыс руб, больше всего предложений работы в Москве и Петербурге.
Так же требуюся специалисты со стажем от 1го до 3х лет, я смею предположить что экспертов и проффесионалов(тех у кого опыт работы более 6ти лет) требуется меньше так как спрос на рынке небольшой и достаточного одного эксперта в команде, в то время как сотрудников с меньшим опытом нужно больше по  нескольким причинам, во первых зарплатные ожидания у них меньше, во вторых они моложе и продуктивность относительно больше хоть и не для не очень сложных задач

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [134]:
# текст запроса
query_5_1 = f'''select e.name, count(v.name)
                from employers as e join vacancies as v on e.id=v.employer_id
                group by 1 
                order by 2 desc
                limit 5
'''

In [135]:
# результат запроса
employer_cnt=pd.read_sql_query(query_5_1,connection)
employer_cnt

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\3135032823.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_cnt=pd.read_sql_query(query_5_1,connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [116]:
# текст запроса

query_5_2 = f'''select a.name,  count(v.name) as vacancies_count, count(e.name) as employer_cnt
                from areas a left join vacancies v  on a.id=v.area_id
                             left join employers e on a.id=e.area
                group by 1
                having count(v.name)=0
                order by 3 desc
'''

In [117]:
# результат запроса
employer_max=pd.read_sql_query(query_5_2,connection)
employer_max

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\2758309799.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_max=pd.read_sql_query(query_5_2,connection)


name  vacancies_count  employer_cnt
0                          Россия                0           410
1                       Казахстан                0           207
2              Московская область                0            75
3              Краснодарский край                0            19
4              Ростовская область                0            18
..                            ...              ...           ...
588                      Струнино                0             0
589    Славгород (Алтайский край)                0             0
590                     Сорочинск                0             0
591  Ленинский (Тульская область)                0             0
592                         Тавда                0             0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [12]:
# текст запроса
query_5_3 = f'''select e.name as employer, count( distinct a.name)
                from employers e join vacancies v on e.id=v.employer_id
                                 join areas a on a.id=v.area_id
                group by 1
                order by 2 desc
''' 

In [13]:
# результат запроса
region_max=pd.read_sql_query(query_5_3,connection)
region_max

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\1395617730.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  region_max=pd.read_sql_query(query_5_3,connection)


employer  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [120]:
# текст запроса
query_5_4 = f'''select count(*)
                from employers e left  join employers_industries ei on e.id=ei.employer_id
                                 left join industries i on  ei.industry_id=i.id
                where i.name is null 
''' 

In [121]:
# результат запроса
null_industry=pd.read_sql_query(query_5_4,connection)
null_industry

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\499503474.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  null_industry=pd.read_sql_query(query_5_4,connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [122]:
# текст запроса

query_5_5 = f'''select e.name, count(i.name)
                from employers e left  join employers_industries ei on e.id=ei.employer_id
                                 left join industries i on  ei.industry_id=i.id
                group by 1
                having count(i.name)=4
                order by 1
                limit 3
''' 

In [123]:
# результат запроса
fourth_employer=pd.read_sql_query(query_5_5,connection)
fourth_employer

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\3999642797.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  fourth_employer=pd.read_sql_query(query_5_5,connection)


name  count
0  101 Интернет      4
1      21vek.by      4
2          2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [124]:
# текст запроса
query_5_6 = f'''select count(*)
                from employers e left  join employers_industries ei on e.id=ei.employer_id
                                 left join industries i on  ei.industry_id=i.id
                where i.name='Разработка программного обеспечения' 
''' 

In [125]:
# результат запроса
employers_developers=pd.read_sql_query(query_5_6,connection)
employers_developers

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\2499549994.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_developers=pd.read_sql_query(query_5_6,connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [126]:
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

In [127]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
links = page.find_all('a') # Ищем все ссылки на странице и сохраняем в переменной links в виде списка
cities_list=[link.text for link in links[46:62]]#сохраняем названия городов миллионеров в список
print(cities_list)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [132]:
# текст запроса
query_5_7 = f'''select  a.name, count(v.id) as total
                from vacancies v join employers e on e.id=v.employer_id
                                 join areas a on a.id=v.area_id
                where e.name='Яндекс' and a.name in ('Москва','Санкт-Петербург','Новосибирск',
                'Екатеринбург','Казань','Нижний Новгород','Красноярск','Челябинск','Самара',
                'Уфа','Ростов-на-Дону','Краснодар','Омск','Воронеж','Пермь','Волгоград')
                group by 1
                union all
                select 'Total' , count(v.id)
                from vacancies v join employers e on e.id=v.employer_id
                                 join areas a on a.id=v.area_id
                where e.name='Яндекс' and a.name in ('Москва','Санкт-Петербург','Новосибирск','Екатеринбург',
                'Казань','Нижний Новгород','Красноярск','Челябинск','Самара','Уфа','Ростов-на-Дону','Краснодар',
                'Омск','Воронеж','Пермь','Волгоград')
                order by 1 desc
''' 

In [133]:
# результат запроса
yandex_vac=pd.read_sql_query(query_5_7,connection)
yandex_vac

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\58371034.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_vac=pd.read_sql_query(query_5_7,connection)


name  total
0         Челябинск     22
1               Уфа     26
2   Санкт-Петербург     42
3            Самара     26
4    Ростов-на-Дону     25
5             Пермь     25
6              Омск     21
7       Новосибирск     35
8   Нижний Новгород     36
9            Москва     54
10       Красноярск     23
11        Краснодар     30
12           Казань     25
13     Екатеринбург     39
14          Воронеж     32
15        Волгоград     24
16            Total    485

***

# выводы по анализу работодателей

Как видим из результатов анализа, самым большим работодателем является Яндекс, по мимо этого есть узнали информацию о том что у 410 работодателей нет вакансий в регионе Россия это может означать что эти работодатели либо не указали регион, так как ищут удвленных сотрудников, где местоположение не важно, либо просто не работают в России.

Так же можно сделать вывод что Яндекс ищет сотрудников в большом колличестве регионов, так же как и Сберанк и Ростелеком, \то говорит что о том что фирмы представлены в этих регонах и сами компании очень большие по численности сотрудников.

из запроса видно что очень много вакансий в городах миллиониках, что в общем то логично, я думаю что такое распредение характертно не тольео для компании Яндекс

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [136]:
# текст запроса
query_6_1 = f'''select count(*)
                from vacancies v
                where lower(v.name) like  lower('%данн%') or lower(v.name) like  lower('%data%')
''' 

In [137]:
# результат запроса
data_vac=pd.read_sql_query(query_6_1,connection)
data_vac

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\451934285.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vac=pd.read_sql_query(query_6_1,connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [139]:
# текст запроса
query_6_2 = f'''select count(*) as filtered_vac
                from vacancies v
                where (lower(v.name) like lower('%data scientist%') 
                       or lower(v.name) like lower('%data science%') 
                       or lower(v.name) like lower('%исследователь данных%') 
                       or lower(v.name) like lower('%машинн%обучен%%')
                       or lower(v.name) like lower('%machine learning%') 
                       or (lower(v.name) like lower('%ML%') and lower(v.name)  not like lower('%HTML%')))
                       and  (v.experience='Нет опыта' 
                       or v.employment='Стажировка'
                       or lower(v.name) like lower('%junior%'))
''' 

In [140]:
# результат запроса
filtered_vac=pd.read_sql_query(query_6_2,connection)
filtered_vac

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\1817396320.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  filtered_vac=pd.read_sql_query(query_6_2,connection)


filtered_vac
0            51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [141]:
# текст запроса
query_6_3 = f'''select count(*) as filtered_vac_sql
                from vacancies v
                where (lower(v.name) like lower('%data scientist%') 
                       or lower(v.name) like lower('%data science%') 
                       or lower(v.name) like lower('%исследователь данных%') 
                       or lower(v.name) like lower('%машинн%обучен%%')
                       or lower(v.name) like lower('%machine learning%') 
                       or (lower(v.name) like lower('%ML%') and lower(v.name)  not like lower('%HTML%')))
                       and (lower(v.key_skills) like lower('%sql%') or lower(v.key_skills) like lower('%postgres%'))
''' 

In [142]:
# результат запроса
filtered_vac_sql=pd.read_sql_query(query_6_3,connection)
filtered_vac_sql

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\4092421078.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  filtered_vac_sql=pd.read_sql_query(query_6_3,connection)


filtered_vac_sql
0               229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [143]:
# текст запроса
query_6_4 = f'''select count(*) as filtered_vac_python
                from vacancies v
                where (lower(v.name) like lower('%data scientist%') 
                       or lower(v.name) like lower('%data science%') 
                       or lower(v.name) like lower('%исследователь данных%') 
                       or lower(v.name) like lower('%машинн%обучен%%')
                       or lower(v.name) like lower('%machine learning%') 
                       or (lower(v.name) like lower('%ML%') and lower(v.name)  not like lower('%HTML%')))
                       and (lower(v.key_skills) like lower('%python%'))
''' 

In [144]:
# результат запроса
filtered_vac_python=pd.read_sql_query(query_6_4,connection)
filtered_vac_python

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\3878332939.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  filtered_vac_python=pd.read_sql_query(query_6_4,connection)


filtered_vac_python
0                  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [145]:
# текст запроса
#зд
query_6_5 = f'''
SELECT 
    round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2) as result
FROM public.VACANCIES v
where 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
'''

In [146]:
# результат запроса
# результат запроса
df = pd.read_sql_query (query_6_5, connection)
df

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\90730344.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_5, connection)


result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [147]:
# текст запроса
query_6_6 = f'''
SELECT 
    experience as "опыт работы",
    ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as "средний размер зарплаты"
FROM public.VACANCIES v
where 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
AND experience = 'От 3 до 6 лет'
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
'''

In [148]:
# результат запроса
df = pd.read_sql_query (query_6_6, connection)
df

C:\Users\npvel\AppData\Local\Temp\ipykernel_2608\3310421505.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_6, connection)


опыт работы  средний размер зарплаты
0  От 3 до 6 лет                 243115.0

***

# выводы по предметному анализу
1.Среди представленных вакансий 1771 имеет отношение к данным

2.Вакансий для новичков среди Data-Science related - 51 штука

3.SQL требуется в 229 вакансиях, Python  в 357 , но это не означает что если язык не указан в требованиях вакансии то он не будет преимуществом при приеме на работу

4.В среднем на интерисующие позиции  которые имеют отношение к анализу данных требуется обладавать 6-7 нвыками

5.Средний разммер зарплаты для специалиста Анализа данных или Data Science  c опытом работы 3-6 лет составляет около 240тыс рублей.


# Общий вывод по проекту

# подведем итог исследования, обобщите выводы

Нам были представлены данные по вакансиям размещенным на hh, в ходе анализа мы узнали что в данных около 50тыс вакансий и около 23тыс раотодателей, не все вакансии имеют указанную зарплату, большинство вакансий  расчитано но полный рабочий день, и требуются в основном специалисты с опытом работы от 1 до 3 лет, в городах миллионниках.

Очень большое количество вакансий  у компаний Яндекс, Сбербанк и Ростелеком и у других крупных компаний что представлены в большом количестве регионов.

из данных можно извлечь много полезной ифнормации о количесве вакансий по опыту раоты, по региону по типу сферы деятельности и многое другое.

Например можно провести исследование зарплат, сравнить их с предлоениями из разных регионов и сделать выводы о наиболее платежеспособном регионе,

или сравнить зарплаты в одном регоне среди сфер деятельности

так же можно определить зависимости сфер деятельности компаний от регионов и определит в каком регионе какая сфера деятельности востребована больше всего.


In [ ]:
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования